### ISSUE TITLE
can not get the relative error which is less than e-3.

### ISSUE THREAD

**Issue by ge1mina023** [mantasu/cs231n#5]
I am learning the Transformer_Captioning.ipynb in assignment3. After I run the cell of testing MultiHeadAttention, I get some incorrect results:
```
self_attn_output error:  0.449382070034207
masked_self_attn_output error:  1.0
attn_output error:  1.0
```
I even copied your MultiHeadAttention code. But, I still get the same result:
```
self_attn_output error:  0.449382070034207
masked_self_attn_output error:  1.0
attn_output error:  1.0
```
Is there anything else I missed?

**Comment by mantasu** [comment]
I am not sure why you get these values, I just tried running a fresh copy of this repository on Google Colab and everything worked fine. It could be a problem with the environment. Can you try cloning a fresh repository and then run `Transformer_Captioning.ipynb` (with only changes to the first code cell)? Or maybe you can try running in Colab?

**Comment by ge1mina023** [comment]
> I am not sure why you get these values, I just tried running a fresh copy of this repository on Google Colab and everything worked fine. It could be a problem with the environment. Can you try cloning a fresh repository and then run `Transformer_Captioning.ipynb` (with only changes to the first code cell)? Or maybe you can try running in Colab?

Yeah. That what I did(with only changes to the first code cell) locally. I guess It's environment issue. My configuration: MacBook Pro M1、Python 3.8.12、conda 4.11.0、torch 2.1.0.dev20230415、macOS Monterey Version 12.5

**Comment by mantasu** [comment]
Maybe you can try Python 3.10 and PyTorch 2.0? Colab currently uses 3.9.16 and 2.0.0

**Comment by mantasu** [comment]
I see, I will try running this in my local environment and I'll let you know. A temporary workaround for now would be to run it on Colab

**Comment by ge1mina023** [comment]
> I see, I will try running this in my local environment and I'll let you know. A temporary workaround for now would be to run it on Colab

Because of some policy reasons, Colab is not convenient for me~

**Comment by mantasu** [comment]
I tried in my local environment and still everything seems fine. Can you reproduce the following:
1. Clone a _completely fresh_ repository:
    ```bash
    git clone https://github.com/mantasu/cs231n
    ```
2. Install PyTorch as described [here](https://pytorch.org/get-started/locally/)
3. Install further requirements:
    ```bash
    pip install h5py numpy matplotlib imageio ipykernel
    ```
4. Change the first code cell in `assignment3/Transformer_Captioning.ipynb` to
    ```bash
    %cd cs231n/datasets/
    !bash get_datasets.sh
    %cd ../../
    ```
5. Run the cells

**Comment by ge1mina023** [comment]
> I tried in my local environment and still everything seems fine. Can you reproduce the following:
> 
> 1. Clone a _completely fresh_ repository:
>    ```shell
>    git clone https://github.com/mantasu/cs231n
>    ```
> 2. Install PyTorch as described [here](https://pytorch.org/get-started/locally/)
> 3. Install further requirements:
>    ```shell
>    pip install h5py numpy matplotlib imageio ipykernel
>    ```
> 4. Change the first code cell in `assignment3/Transformer_Captioning.ipynb` to
>    ```shell
>    %cd cs231n/datasets/
>    !bash get_datasets.sh
>    %cd ../../
>    ```
> 5. Run the cells

I have finished all steps of yours(only Jupyter was installed extra), and I have python 3.9.16 and PyTorch 2.0.
But, it's still doesn't work.
I still got
```
self_attn_output error:  0.449382070034207
masked_self_attn_output error:  1.0
attn_output error:  1.0
```
in my first computable cell.

**Comment by mantasu** [comment]
Hmm, do you only get that in `Transformer_Captioning.ipynb`? Also, have you tried other solutions (e.g., from other repositories)?

**Comment by ge1mina023** [comment]
> Hmm, do you only get that in `Transformer_Captioning.ipynb`? Also, have you tried other solutions (e.g., from other repositories)?

Yeah, I have tried other solutions. And Only Transformer_Captioning.ipynb has this error in assignment3.

**Comment by mantasu** [comment]
If even other solutions don't work, it's most likely a problem with the environment. I would suggest trying the following:
* Reinstall Python and Conda
* Running on another device
* Running on a server, e.g., Kaggle, Azure Notebook, SageMaker, CoCalc. I would still suggest using Colab, e.g., with VPN
* Running on a virtual machine

**Comment by ge1mina023** [comment]
> If even other solutions don't work, it's most likely a problem with the environment. I would suggest trying the following:
> 
> * Reinstall Python and Conda
> * Running on another device
> * Running on a server, e.g., Kaggle, Azure Notebook, SageMaker, CoCalc. I would still suggest using Colab, e.g., with VPN
> * Running on a virtual machine

Thanks, I will try.

**Comment by xiaoyatang** [comment]
> > If even other solutions don't work, it's most likely a problem with the environment. I would suggest trying the following:
> > 
> > * Reinstall Python and Conda
> > * Running on another device
> > * Running on a server, e.g., Kaggle, Azure Notebook, SageMaker, CoCalc. I would still suggest using Colab, e.g., with VPN
> > * Running on a virtual machine
> 
> Thanks, I will try.
Hi, have you solved the issue? i just met the same one and solved. I would say your error is due to calculation itself rather than environment. It is not like an issue of computational precision. My suggestion is that make sure you follow the right order , query*key->mask->softmax->attn_dropout-> attn*value->projection.

**Comment by putskan** [comment]
> > > If even other solutions don't work, it's most likely a problem with the environment. I would suggest trying the following:
> > > 
> > > * Reinstall Python and Conda
> > > * Running on another device
> > > * Running on a server, e.g., Kaggle, Azure Notebook, SageMaker, CoCalc. I would still suggest using Colab, e.g., with VPN
> > > * Running on a virtual machine
> > 
> > 
> > Thanks, I will try.
> > Hi, have you solved the issue? i just met the same one and solved. I would say your error is due to calculation itself rather than environment. It is not like an issue of computational precision. My suggestion is that make sure you follow the right order , query_key->mask->softmax->attn_dropout-> attn_value->projection.

Had the same problem with my solution, and using this repo's solution also gave me the exact same error.
Are you sure it's a calculation error?

Also using M1. I don't think it's a coincidence.

**Comment by PeterHUistyping** [comment]
Also having such problem with Mac M1 silicon. 

By debugging the individual lines of code, one potential reason is the `nn.Dropout()`, also reported at [Dropout isn't deterministic between Intel and M1 MacBooks](https://github.com/pytorch/pytorch/issues/84030). 
> completely reproducible results are not guaranteed across PyTorch releases, individual commits, or different platforms. Furthermore, results may not be reproducible between CPU and GPU executions, even when using identical seeds.

**Comment by nitin0497** [comment]
> I am learning the Transformer_Captioning.ipynb in assignment3. After I run the cell of testing MultiHeadAttention, I get some incorrect results:
> 
> ```
> self_attn_output error:  0.449382070034207
> masked_self_attn_output error:  1.0
> attn_output error:  1.0
> ```
> 
> I even copied your MultiHeadAttention code. But, I still get the same result:
> 
> ```
> self_attn_output error:  0.449382070034207
> masked_self_attn_output error:  1.0
> attn_output error:  1.0
> ```
> 
> I even downloaded your assignment3 code, and I still get the same output.
> 
> Is there anything else I missed?

I have been struggling with it for a few hours now, and I have finally found the solution that works for me. I hope it works for you, too. Simply add the following code to a new cell in your colab notebook Transformer_Captioning.ipynb 


!pip uninstall -y torch torchvision torchaudio
!pip install --upgrade pip
!pip install --index-url https://download.pytorch.org/whl/cu121 torch torchvision torchaudio

import os, sys
os.kill(os.getpid(), 9)

**Comment by willcel** [comment]
> > I am learning the Transformer_Captioning.ipynb in assignment3. After I run the cell of testing MultiHeadAttention, I get some incorrect results:
> > ```
> > self_attn_output error:  0.449382070034207
> > masked_self_attn_output error:  1.0
> > attn_output error:  1.0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I even copied your MultiHeadAttention code. But, I still get the same result:
> > ```
> > self_attn_output error:  0.449382070034207
> > masked_self_attn_output error:  1.0
> > attn_output error:  1.0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I even downloaded your assignment3 code, and I still get the same output.
> > Is there anything else I missed?
> 
> I have been struggling with it for a few hours now, and I have finally found the solution that works for me. I hope it works for you, too. Simply add the following code to a new cell in your colab notebook Transformer_Captioning.ipynb
> 
> !pip uninstall -y torch torchvision torchaudio !pip install --upgrade pip !pip install --index-url https://download.pytorch.org/whl/cu121 torch torchvision torchaudio
> 
> import os, sys os.kill(os.getpid(), 9)

That works for me.
Sure this issue is related to "randn seed" difference between torch versions.

Thanks, Time-saver

**Comment by nitin0497** [comment]
@willcel, you are welcome. Someone suggested using the 2025.07 version of Python in Colab instead of the latest, which is also a simple fix. I haven't tried, though.

**Comment by willcel** [comment]
> [@willcel](https://github.com/willcel), you are welcome. Someone suggested using the 2025.07 version of Python in Colab instead of the latest, which is also a simple fix. I haven't tried, though.

I tried that (2025.07 version of colab runtime). Works. 
Detail: 
In colab, press Ctrl + Shift + P (Windows), then search: fallback. Switch Colab Runtime "2025.07" restart.

Even better than "!pip install.." solution, because pip install throw error when i change to another assignment colab (don't know why)
